### PR TITLE
feat: evaluate HistFactory constraints in log space

### PR DESCRIPTION
**Pull request by kratsg**
:robot: _AI text below_ :robot:

## Summary

Layer 2 of #243: no site computes `pt.log(<probability-space constraint>)` anymore. `HasConstraint` (and the channel-level BB-lite helper) now expose a `log_constraint` counterpart to `make_constraint`, built from the *same* constraint distribution object(s) via their analytic `log_likelihood`/`log_expression` (from #261) instead of round-tripping through a probability value.

- **`src/pyhs3/distributions/histfactory/modifiers.py`**: `HasConstraint.log_constraint()` is now abstract alongside `make_constraint()`. Each modifier factors its distribution construction into a shared private helper consumed by both:
  - `SingleParamConstraint` (`NormSys`/`HistoSys`): `_build_constraint()` builds the single Gauss/Poisson/LogNormal distribution once; `make_constraint`/`log_constraint` call `expression()`/`log_expression()` on it.
  - `ShapeSysModifier`: `_build_bin_constraints()` builds the per-bin Poisson distributions once; `make_constraint` takes their product, `log_constraint` sums their log-probabilities.
  - `StatErrorModifier` (BB-full): same per-bin sharing, for whichever of Gauss/Poisson the modifier is configured with.
- **`src/pyhs3/distributions/histfactory/__init__.py`**: the channel-level BB-lite constraint gets the same treatment — `_build_barlow_beeston_lite_dists()` is shared by `_make_barlow_beeston_lite_constraint()` and the new `_make_barlow_beeston_lite_log_constraint()`. `HistFactoryDistChannel.log_extended_likelihood` now sums `modifier.log_constraint(...)` and the BB-lite log constraint directly, never `pt.log(make_constraint(...))`.
- **`src/pyhs3/model.py`**: `Model._build_distribution_node`'s single pass over constraint specs now calls `modifier.log_constraint(...)` alongside `modifier.make_constraint(...)` (and the BB-lite log helper alongside the prob one), building each constraint's log-space subgraph exactly once per spec. A new `self._hfdc_log_constraints` cache mirrors `self._hfdc_constraints` with the same cross-channel dedup semantics; `Model.log_prob` extends its terms with `self._hfdc_log_constraints` directly instead of `pt.log(self._hfdc_constraints)`, and the per-channel `log_distributions` entry is assembled from `channel_log_constraints` instead of `pt.log(channel_constraints)`.

No normalization-quadrature complication was found in the constraint path: `SingleParamConstraint`/`ShapeSysModifier`/`StatErrorModifier`/BB-lite all build a fresh `Context` from a plain dict (losing the outer `observables` dict), so `Distribution._apply_normalization` never applies to these constraint distributions and `log_expression() == log_likelihood()` exactly for all of them — `log_constraint` is a clean analytic mirror of `make_constraint` everywhere.

Fixes #243
Refs #254 (phase 2)

## Acceptance evidence

For a single-bin HFDC channel with one `normsys` (Gaussian) constraint, pushed to `alpha_lumi = 40.0` (probability-space Gaussian constraint underflows to exactly `0.0` — `exp(-40**2/2) < 2.2e-308`):

- `model.log_prob` (compiled): **finite**, `-1207.4659839363708`
- `logpdf("SR", alpha_lumi=40.0, ...)`: **finite**, `-1207.4659839363708`
- Analytic reference (Poisson log-pmf + `scipy.stats.norm.logpdf(40.0)`): `-1207.4659839363708` (Poisson term `-406.5470454031661` + Gaussian term `-800.9189385332047`)
- Both match to `rel=1e-8` in the added regression tests.

For the bare `normsys` modifier constraint at `alpha_lumi = 40.0`:
- `make_constraint(...)` → `0.0` (probability-space underflow)
- `log_constraint(...)` → `-800.9189385332047`, matching `scipy.stats.norm.logpdf(40.0)` to `rel=1e-10`

Parity (`log_constraint == log(make_constraint)`, `rtol` 1e-8 to 1e-10) verified at ordinary parameter values for every constraint-bearing modifier: `NormSysModifier`/`HistoSysModifier` (Gauss/Poisson/LogNormal), `ShapeSysModifier`, `StatErrorModifier` (Gauss/Poisson, including the zero-yield-bin skip and empty-parameters edge cases), and the channel-level BB-lite constraint (Gauss/Poisson).

## Test plan

- [x] New tests added first (TDD): `tests/test_histfactory_modifiers.py::TestLogConstraintParity`/`TestLogConstraintUnderflow`, `tests/test_hfdc_model_constraints.py::TestBBLiteStaterrorModelBuild::test_bblite_log_constraint_matches_log_of_constraint` and `TestHFDCSubgraphBuildCounts::test_log_constraint_called_once_per_spec`, `tests/test_log_space_likelihood.py::test_log_prob_finite_for_normsys_constraint_at_alpha_40`/`test_logpdf_finite_for_normsys_constraint_at_alpha_40`
- [x] Full quick suite green: `pixi run --environment py312 test` (1100 passed, 3 skipped, 1 xfailed)
- [x] Coverage 100% lines+branches on all touched files (`modifiers.py`, `distributions/histfactory/__init__.py`; `model.py`'s only miss is the pre-existing, unrelated `--runpydot`-gated graph-visualization method)
- [x] `pre-commit run --files <touched files>` clean
- [x] Slow real-world suite: `pixi run --environment py312 test-slow tests/test_realworld.py` (5 passed, 1 xfailed, ~2m9s)
- [x] No existing tolerance loosened

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved likelihood calculations to stay finite when constraint probabilities become extremely small.
  * Fixed constraint handling so log-space results now match probability-space results across several modifier types.
  * Added safeguards for zero-yield bins and empty constraint cases.

* **Tests**
  * Added regression coverage for log-space constraint consistency and underflow behavior.
  * Expanded checks for HistFactory model constraint construction and log-likelihood calculations.

### CODE REVIEW

File: src/pyhs3/distributions/histfactory/modifiers.py [hunk 1]
```diff
@@ -351,21 +383,28 @@ def make_constraint(self, context: Context, sample_data: SampleData) -> TensorVa
...
+        pairs = self._build_bin_constraints(context, sample_data)
+        factors = [dist.expression(ctx) for dist, ctx in pairs]
         return cast(TensorVar, pt.prod(pt.stack(factors), axis=0))  # type: ignore[no-untyped-call]
 
+    def log_constraint(self, context: Context, sample_data: SampleData) -> TensorVar:
+        """Create constraint term as the sum of per-bin Poisson log-probabilities."""
+        pairs = self._build_bin_constraints(context, sample_data)
+        log_terms = [dist.log_expression(ctx) for dist, ctx in pairs]
+        return cast(TensorVar, pt.sum(pt.stack(log_terms), axis=0))  # type: ignore[no-untyped-call]
+
```
[the remark this answers is not in the document]
Comment (kratsg) [reply src/pyhs3/distributions/histfactory/modifiers.py:407]: self.parameters cannot be empty.
